base/src/main/java/org/aya/core/repr/AyaShape.java

```diff
@@ -18,29 +18,40 @@
  */
 public sealed interface AyaShape {
   @NotNull CodeShape codeShape();
-  @NotNull Term transform(@NotNull Term term, @NotNull Term type);
-
-  @NotNull CodeShape DATA_NAT = new DataShape(ImmutableSeq.empty(), ImmutableSeq.of(
-    new CtorShape(ImmutableSeq.empty()),
-    new CtorShape(ImmutableSeq.of(CodeShape.ParamShape.ex(new CodeShape.TermShape.Call(0))))
-  ));
 
   @NotNull AyaShape NAT_SHAPE = new AyaIntLitShape();
   @NotNull ImmutableSeq<AyaShape> LITERAL_SHAPES = ImmutableSeq.of(NAT_SHAPE);
 
   record AyaIntLitShape() implements AyaShape {
+    public static final @NotNull CodeShape DATA_NAT = new DataShape(ImmutableSeq.empty(), ImmutableSeq.of(
+      new CtorShape(ImmutableSeq.empty()),
+      new CtorShape(ImmutableSeq.of(CodeShape.ParamShape.ex(new CodeShape.TermShape.Call(0))))
+    ));
+
     @Override public @NotNull CodeShape codeShape() {
       return DATA_NAT;
     }
-
-    @Override public @NotNull Term transform(@NotNull Term term, @NotNull Term type) {
-      return term;
-    }
   }
 
-  record Factory(@NotNull MutableMap<Def, MutableList<AyaShape>> discovered) {
+  record Factory(@NotNull MutableMap<Def, AyaShape> discovered) {
     public Factory() {
       this(MutableLinkedHashMap.of());
     }
+
+    public @NotNull ImmutableSeq<Def> findImpl(@NotNull AyaShape shape) {
+      return discovered.view().map(Tuple::of)
+        .filter(t -> t._2 == shape)
+        .map(t -> t._1)
+        .toImmutableSeq();
+    }
+
+    public @NotNull Option<AyaShape> find(@NotNull Def def) {
+      return discovered.getOption(def);
+    }
+
+    public void bonjour(@NotNull Def def, @NotNull AyaShape shape) {
+      // TODO[literal]: what if a def has multiple shapes?
```

Comment:
For now I believe there cannot be multiple shapes. I can only think of integers and lists for now.

Reply:
Then let's be a TODO.